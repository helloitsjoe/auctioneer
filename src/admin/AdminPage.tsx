import * as React from 'react';
import {connect} from 'react-redux';
import {submitChange, deleteRequest } from '../actions/adminActions';
import {AuctionItem, Modal, selectAuctionItems} from '../reducers';

import { ItemEditor, InputKey } from './ItemEditor';
import { Poller } from '../Poller';
import { Sidebar } from './Sidebar';
import { AdminHeader } from './AdminHeader';
import { ConfirmDiscard } from './ConfirmDiscard';
import { MissingInfoNotice } from './MissingInfoNotice';
import { createNewAuctionItem } from '../utils';

type StoreProps = {
    initialItems: AuctionItem[];
}

type DispatchProps = {
    submitChange: (item: AuctionItem) => void;
    deleteRequest: (id: number) => void;
}

type Props = StoreProps & DispatchProps & {
    poller: Poller;
    initialItems: AuctionItem[];
}

type State = {
    items: AuctionItem[];
    currentItem: AuctionItem;
    dirty: boolean;
    focusedIndex: number;
    missingInfo: boolean;
    confirmDiscard: boolean;
}

export class AdminPage extends React.Component<Props, State> {

    static defaultProps = {
        poller: new Poller()
    }

    state = {
        items: this.props.initialItems,
        dirty: false,
        focusedIndex: 0,
        currentItem: this.props.initialItems[0],
        missingInfo: false,
        confirmDiscard: false,
    }

    componentDidMount() {
        this.props.poller.stop();
    }

    handleInputChange = (inputType: InputKey) => e => {
        const { value } = e.target;
        this.setState(prevState => {
            const { currentItem } = prevState;
            const bids = currentItem.bids.map(bid =>
                bid.name === 'min' ? { ...bid, value: parseInt(value) } : bid)
            const inputMap = {
                title: { title: value },
                description: { description: value },
                minBid: { bids }
            }
            const newItem = { ...currentItem, ...inputMap[inputType] };
            return { dirty: true, missingInfo: false, currentItem: newItem }
        });
    }

    handleItemFocus = (i: number) => {
        if (this.state.dirty) {
            return this.setState({ confirmDiscard: true });
        }
        this.setState(prevState => {
            return {
                focusedIndex: i,
                currentItem: prevState.items[i],
            };
        });
    }

    handleAddItem = () => {
        this.setState(prevState => {
            const { items } = prevState;
            const blankItemIndex = items.findIndex(item => !item.title.length);
            if (blankItemIndex > -1) {
                return {
                    items,
                    focusedIndex: blankItemIndex,
                    currentItem: items[blankItemIndex],
                }
            }
            const newItem = createNewAuctionItem(items);
            const itemsWithNew = [...items, newItem];
            return {
                items: itemsWithNew,
                currentItem: newItem,
                focusedIndex: itemsWithNew.indexOf(newItem)
            };
        });
    }

    handleSaveChanges = (focusedItem: AuctionItem) => e => {
        this.props.submitChange(focusedItem);
        this.setState({ dirty: false, confirmDiscard: false });
    }

    handleDiscardChanges = e => {
        this.setState(prevState => ({
            currentItem: prevState.items[prevState.focusedIndex],
            dirty: false,
            confirmDiscard: false
        }));
    }

    handleSubmit = (e: any) => {
        e.preventDefault();
        const { currentItem } = this.state;
        if (!(currentItem.title.trim() && currentItem.description.trim())) {
            return this.setState({ missingInfo: true });
        }
        this.props.submitChange(currentItem);
        this.setState({ dirty: false, missingInfo: false });
    }

    handleDelete = (e: any) => {
        this.setState(prevState => {
            const { items, focusedIndex} = prevState;
            const focusedItem = items[focusedIndex];
            this.props.deleteRequest(focusedItem.id);
            // TODO: Don't filter both here and in reducer
            const filteredItems = items.filter(({id}) => id !== focusedItem.id);
            const safeItemsAfterDelete = filteredItems.length ? filteredItems : [createNewAuctionItem()];
            const safeFocusedIndex = focusedIndex >= safeItemsAfterDelete.length
                ? safeItemsAfterDelete.length - 1
                : focusedIndex;
            const safeItemAfterDelete = safeItemsAfterDelete[safeFocusedIndex];
            return {
                focusedIndex: safeFocusedIndex,
                items: safeItemsAfterDelete,
                currentItem: safeItemAfterDelete,
            }
        });
    }

    closeModal = (name: Modal) => e => {
        if (!e.key || e.key === 'Escape') {
            switch (name) {
                case Modal.missingInfo:
                    return this.setState({ missingInfo: false });
                case Modal.confirmDiscard:
                    return this.setState({ confirmDiscard: false });
            }
        }
    }

    render() {
        const {
            items,
            currentItem,
            focusedIndex,
            missingInfo,
            confirmDiscard
        } = this.state;

        return (
            <div>
                {confirmDiscard &&
                    <ConfirmDiscard
                        onCloseModal={this.closeModal(Modal.confirmDiscard)}
                        onSaveChanges={this.handleSaveChanges(currentItem)}
                        onDiscardChanges={this.handleDiscardChanges}
                    />
                }
                {missingInfo &&
                    <MissingInfoNotice
                        onCloseModal={this.closeModal(Modal.missingInfo)}
                    />
                }
                <AdminHeader />
                <div className="admin-page">
                    <Sidebar
                        items={items}
                        focusedItem={currentItem}
                        focusedIndex={focusedIndex}
                        onAddItem={this.handleAddItem}
                        onItemFocus={this.handleItemFocus}
                    />
                    <ItemEditor
                        item={currentItem}
                        onSubmit={this.handleSubmit}
                        onDelete={this.handleDelete}
                        onChangeTitle={this.handleInputChange(InputKey.title)}
                        onChangeMinBid={this.handleInputChange(InputKey.minBid)}
                        onChangeDescription={this.handleInputChange(InputKey.description)}
                    />
                </div>
            </div>
        )
    }
}

const mStP = (state): StoreProps => ({ initialItems: selectAuctionItems(state) });
const mDtP: DispatchProps = { submitChange, deleteRequest };

export default connect(mStP, mDtP)(AdminPage);