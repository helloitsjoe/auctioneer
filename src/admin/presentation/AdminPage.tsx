import * as React from 'react';
import { ItemEditor } from './ItemEditor';
import { Sidebar } from './Sidebar';
import { AdminHeader } from './AdminHeader';

type Props = {
    auctionItems: any;
}

export class AdminPage extends React.Component<Props, any> {

    // TODO: itemClickHandler

    render() {
        return <div>
            <AdminHeader />
            <div className="admin-page">
                <Sidebar auctionItems={this.props.auctionItems}/>
                <ItemEditor />
            </div>
        </div>
    }
}
