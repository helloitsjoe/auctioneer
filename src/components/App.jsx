import * as React from 'react';

// export interface AppProps { compiler:string; framework:string; }

export class App extends React.Component { // App State is undefined
    render() {
        return <h1>Hello from {this.props.compiler} and {this.props.framework}!</h1>;
    }
}
