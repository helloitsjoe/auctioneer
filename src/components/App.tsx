import * as React from 'react';
import {BrowserRouter as Router, Route } from 'react-router-dom';
import axios from 'axios';
import Nav from './Nav';
import { List } from './List';
import { Footer } from './Footer';
import { UserList } from './UserList';
import { UserNameForm } from './UserNameForm';
import { DATA_URL, randFromArr, DEFAULT_NAMES } from '../utils';

type State = {
    error: string;
    isLoaded: boolean;
    userTotal: number;
    auctionItems: any;
}

export class App extends React.Component<any, State> {

    constructor(props) {
        super(props);

        this.state = {
            error: null,
            isLoaded: false,
            userTotal: 0,
            auctionItems: null,
        };

        window.sessionStorage.userID = window.sessionStorage.userID || randFromArr(DEFAULT_NAMES);
    }

    public async componentDidMount() {
        try {
            const response = await axios.get(DATA_URL);
            const auctionItems = response && response.data;
            // TODO: Don't use sessionStorage to store userID
            console.log('state data:', auctionItems);
            this.setState({ auctionItems, isLoaded: true });
        } catch (err) {
            console.error(err);
            this.setState({ error: JSON.stringify(err), isLoaded: true });
        }
    }

    public render() {
        // I could use a filtering button instead of Routes, just wanted to experiment with React Router
        return !this.state.isLoaded ? <div>Loading...</div>
            : this.state.error ? <div>Error: {this.state.error}</div>
            : (
                <Router>
                    <div className="well">
                        <UserNameForm />
                        <Nav />
                        <Route exact={true} path="/" render={() => <List data={this.state.auctionItems} />} />
                        <Route exact={true} path="/user" render={() => <UserList data={this.state.auctionItems} />} />
                        <Footer userTotal={this.state.userTotal} />
                    </div>
                </Router>
            );
    }
}
