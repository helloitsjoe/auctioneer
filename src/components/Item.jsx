import * as React from 'react';

export class Item extends React.Component {
    render() {
        return (
            <div>
                // Title (pull left)
                // Button Box (pull right)
                    // Current Price - value from state
                    // Bid button
                        // Label: Current price + $5? $10? Just '$5 more'?
                        // Two buttons? '+$5' and 'Custom bid'?
            </div>
        );
    };

    handleClick() {
        // If tap on title, open up description/photos
        // If tap on $5 button, update current bid on DB and in view
    }
}
