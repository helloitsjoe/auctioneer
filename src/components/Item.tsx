import * as React from 'react';
import axios from 'axios';
import { DATA_URL } from '../utils';

type Props = {
    itemData: any;
}

type State = {
    highBid: number;
    highBidder: string;
}

// TODO: Refactor this file into container/presentation components
export default class Item extends React.Component<Props, State> {

    constructor(props) {
        super(props);

        const { highBid, highBidder } = this.getHighBid(this.props.itemData.bids);
        this.state = {
            highBid,
            highBidder
        };

        this.quickBid = this.quickBid.bind(this);
        this.updateData = this.updateData.bind(this);
        this.toggleDescription = this.toggleDescription.bind(this);
    }

    getHighBid(bids) {
        let highBid = 0;
        let highBidder = '';

        for (let j = 0; j < bids.length; j++) {
            if(bids[j].bid > highBid) {
                highBid = bids[j].bid;
                highBidder = bids[j].name;
            }
        }
        return { highBid, highBidder }
    }

    quickBid(e) {
        e.stopPropagation();
        const itemData = this.props.itemData;
        let newBid = this.getHighBid(this.props.itemData.bids).highBid + 5;
        // TODO: Don't user window.sessionStorage
        itemData.bids.push({ name: window.sessionStorage.userID, bid: newBid });
        const { highBid, highBidder } = this.getHighBid(itemData.bids);

        this.setState({ highBid, highBidder });
        this.styleItem(highBidder);
        this.updateData();
    }

    private updateData(): void {
        const body = JSON.stringify(this.props.itemData);
        axios.put(DATA_URL, { body });
    }

    private styleItem(highBidder: string): void {
        const user = window.sessionStorage.userID;
        const currentItem = document.getElementById(`item-${this.props.itemData.id}`);
        const yourBid = document.getElementById(`bid-text-${this.props.itemData.id}`);
        if (highBidder === user) {
            currentItem.classList.add('bid-bg');
            currentItem.classList.remove('outbid-bg');
            yourBid.innerHTML = 'High bid (You!)';
            yourBid.classList.add('user-high-bid');
            yourBid.classList.remove('user-outbid');
        } else if (this.props.itemData.bids.find(item => item.name === user)) {
            currentItem.classList.add('outbid-bg');
            currentItem.classList.remove('bid-bg');
            yourBid.innerHTML = 'High bid (Not you!)';
            yourBid.classList.remove('user-high-bid');
            yourBid.classList.add('user-outbid');  
        }
    }

    toggleDescription(e) {
        e.stopPropagation();
        let description = document.getElementById(`description-${this.props.itemData.id}`);
        description.classList.toggle('open');

        // TODO: Add photos
    }

    componentDidMount() {
        this.styleItem(this.state.highBidder);
    }

    render() {
        return (
            <div className="item-group clearfix" onClick={this.toggleDescription}>
                <div className="item-container" id={`item-${this.props.itemData.id}`}>
                    <div className="item-title u-pull-left">
                        <span>{this.props.itemData.title}</span>
                    </div>
                    <div className="button-box u-pull-right">
                    <span className="bid-text" id={`bid-text-${this.props.itemData.id}`}>High bid:</span>
                        <span className="high-bid" id={`high-bid-${this.props.itemData.id}`}>{this.state.highBid}</span>
                        <button className="bid btn" onClick={this.quickBid}>Bid {this.state.highBid + 5}</button>
                    </div>
                </div>
                <div className="description" id={`description-${this.props.itemData.id}`}>{this.props.itemData.description}</div>
            </div>
        );
    };
}
