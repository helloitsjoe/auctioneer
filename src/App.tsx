import * as React from 'react';
import axios from 'axios';
import { connect } from 'react-redux';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';

import { Poller } from './Poller';
import { BidsPage } from './user/BidsPage';
import { randFromArr, DEFAULT_NAMES, DATA_URL } from './utils';
import { AdminPage } from './admin/AdminPage';
import { setAuctionData, setAuctionError } from './actions/auctionItemActions';
import { ItemData, selectError, selectIsLoaded, selectAuctionItems } from './reducers';

type Props = {
    axios: any
    error: Error,
    isLoaded: boolean,
    auctionItems: ItemData[],
    setAuctionData: (auctionItems: ItemData[], userName: string) => void;
    setAuctionError: (errorMessage: string) => void;
    poller: Poller,
}

export class App extends React.Component<Props> {

    static defaultProps = {
        poller: new Poller(),
        axios: axios,
    }

    constructor(props) {
        super(props);
        const userName = sessionStorage.getItem('userName') || randFromArr(DEFAULT_NAMES).toUpperCase();
        sessionStorage.setItem('userName', userName);
    }

    async componentDidMount() {
        await this.fetchAuctionData();
        // Kick off poll every second for new auction data... TODO: Make this a socket?
        this.props.poller.init(this.fetchAuctionData);
    }

    componentWillUnmount() {
        this.props.poller.stop();
    }

    componentDidCatch(err, info) {
        console.error(err, info);
    }

    fetchAuctionData = async () => {
        // TODO: Move this into a thunk
        try {
            const response = await this.props.axios.get(DATA_URL);
            const auctionItems = response && response.data;
            this.props.setAuctionData(auctionItems, sessionStorage.getItem('userName'));
        } catch (err) {
            console.error(err);
            this.props.setAuctionError(err.message);
            this.props.poller.stop();
        }
    }

    render() {
        const { error, isLoaded, auctionItems } = this.props;

        return !isLoaded ? <div>Loading...</div>
            : error ? <div>Error: {JSON.stringify(error)}</div>
            : (
                <Router>
                    <Switch>
                        <Route exact path="/admin" render={() => 
                            <AdminPage poller={this.props.poller} />} />
                        <Route path="/" render={({location}) => 
                            <BidsPage
                                poller={this.props.poller}
                                auctionItems={auctionItems}
                                user={sessionStorage.getItem('userName')}
                                filter={location.pathname === '/user'} />} />
                    </Switch>
                </Router>
            );
    }
}

const mapStateToProps = state => ({
  error: selectError(state),
  isLoaded: selectIsLoaded(state),
  auctionItems: selectAuctionItems(state)
});

const mapDispatchToProps = {
    setAuctionData,
    setAuctionError
}

export default connect(mapStateToProps, mapDispatchToProps)(App);