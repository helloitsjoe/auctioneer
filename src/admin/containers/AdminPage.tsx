import axios from 'axios';
import * as React from 'react';
import { Sidebar } from '../presentation/Sidebar';
import { ItemEditor } from './ItemEditor';
import { AdminHeader } from '../presentation/AdminHeader';
import { ItemData } from '../../containers/App';
import { DATA_URL, createNewAuctionItem } from '../../utils';

type Props = {
    auctionItems: ItemData[];
}

type State = {
    selectedIndex: number;
    auctionItems: ItemData[];
}

export class AdminPage extends React.Component<Props, State> {

    constructor(props) {
        super(props);

        this.state = {
            selectedIndex: 0,
            auctionItems: this.props.auctionItems,
        }
    }

    handleClick = (i, e) => {
        if (i === null) { // AddItem
            const id = this.state.auctionItems.length;
            const auctionItems = [...this.state.auctionItems, createNewAuctionItem({ id })];
            this.setState({ auctionItems, selectedIndex: id  });
        } else {
            this.setState({ selectedIndex: i });
        }
    }

    updateTitle = (title, id) => {
        // This seems weird. Probably a better way to do this. Redux?
        const auctionItems = this.state.auctionItems.map(item => {
            if (item.id === id) {
                item.title = title;
            }
            return item;
        });
        this.setState({ auctionItems });
    }

    submitChanges = (itemState, e) => {
        e.preventDefault();
        const { auctionItems, selectedIndex } = this.state;

        const selectedItem = auctionItems[selectedIndex];
        selectedItem.bids[0].value = itemState.minBid;
        const updatedItem = Object.assign({}, selectedItem, itemState);

        axios.put(DATA_URL, { body: JSON.stringify(updatedItem) });

        const updatedItems = auctionItems.map(item => (item.id === itemState.id) ? updatedItem : item);

        this.setState({ auctionItems: updatedItems });
    }

    render() {
        return <div>
            <AdminHeader />
            <div className="admin-page">
                <Sidebar
                    clickHandler={this.handleClick}
                    auctionItems={this.state.auctionItems}
                    selectedIndex={this.state.selectedIndex} />
                <ItemEditor
                    updateTitle={this.updateTitle}
                    submitChanges={this.submitChanges}
                    itemData={this.state.auctionItems[this.state.selectedIndex]} />
            </div>
        </div>
    }
}
