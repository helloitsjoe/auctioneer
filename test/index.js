describe('Auctioneer', function () {
    require('./redux.test');
    require('./utils.test');
    require('./App.test');
    require('./Item.test');
    require('./List.test');
    require('./Server.test');
    require('./Poller.test');
    require('./BidsPage.test');
    require('./AdminPage.test');
});