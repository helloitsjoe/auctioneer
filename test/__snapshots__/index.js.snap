// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auctioneer redux duck tests User load auction data 1`] = `
Array [
  Object {
    "bids": Array [
      Object {
        "name": "Me",
        "value": 75,
      },
    ],
    "description": "Hello",
    "id": 45,
    "title": "Boo",
    "viewDetails": false,
  },
]
`;

exports[`Auctioneer redux duck tests User load auction data 2`] = `
Array [
  Object {
    "bids": Array [
      Object {
        "name": "Me",
        "value": 75,
      },
    ],
    "description": "Hello",
    "id": 45,
    "title": "Boo",
    "viewDetails": false,
  },
  Object {
    "bids": Array [
      Object {
        "name": "You",
        "value": 75,
      },
    ],
    "description": "Hi again",
    "id": 50,
    "title": "Bee",
    "viewDetails": false,
  },
]
`;
